core/src/main/scala/kafka/network/PrioritizationAwareBlockingQueue.scala

```diff
@@ -0,0 +1,139 @@
+
+/**
+ * A blocking queue that is prioritization-aware.
+ *
+ * Prioritized elements in the queue get polled before regular elements in the queue.
+ */
+private[network] class PrioritizationAwareBlockingQueue[E](private val capacity: Int) {
+  private var count: Int = 0
+  private val prioritizedElements: ArrayDeque[E] = new ArrayDeque[E]()
+  private val regularElements: ArrayDeque[E] = new ArrayDeque[E]()
+  private val lock: ReentrantLock = new ReentrantLock()
+  private val notFull: Condition = lock.newCondition()
+  private val notEmpty: Condition = lock.newCondition()
+
+  /**
+   * Inserts the potentially prioritized element into the queue, blocking if space is unavailable.
+   *
+   * @param e - non-null element to be inserted
+   * @param prioritized - flag indicating whether the element should be prioritized
+   * @throws InterruptedException
+   */
+  def put(e: E, prioritized: Boolean): Unit = {
+    if (e == null) {
+      throw new NullPointerException()
+    }
+    lock.lockInterruptibly()
+    try {
+      while (count == capacity) {
+        notFull.await()
+      }
+      enqueue(e, prioritized)
+    } finally {
+      lock.unlock()
+    }
+  }
+
+  /**
+   * Return and remove elements from the head of the queue in a priority-aware manner, blocking up to the provided
+   * timeout for elements.
+   *
+   * That is, prioritized elements in the queue get polled before regular elements in the queue.
+   *
+   * @param timeout - the maximum amount of time to wait for an item to be polled from the queue
+   * @param timeUnit - the unit of time
+   * @return potentially prioritized element from the head of the queue
+   * @throws InterruptedException
+   */
+  def poll(timeout: Long, timeUnit: TimeUnit): E = {
+    var nanos = timeUnit.toNanos(timeout)
+    lock.lockInterruptibly()
+    try {
+      while (count == 0) {
+        if (nanos <= 0) {
+          return null.asInstanceOf[E]
+        }
+        nanos = notEmpty.awaitNanos(nanos)
+      }
+      dequeue
+    } finally {
+      lock.unlock()
+    }
+  }
+
+  /**
+   * Inserts the potentially prioritized element into the queue.
+   *
+   * Call only within the lock when count < capacity.
+   *
+   * @param e - non-null element to be inserted
+   * @param prioritized - flag indicating whether the element should be prioritized
+   */
+  private def enqueue(e: E, prioritized: Boolean): Unit = {
+    if (prioritized) {
+      prioritizedElements.add(e)
+    } else {
+      regularElements.add(e)
+    }
+    count += 1
+    notEmpty.signal
+  }
+
+  /**
+   * Return and remove elements from the head of the queue in a priority-aware manner.
+   *
+   * Call only within the lock when count > 0.
+   * Since count > 0, it's guaranteed that one of the queues is nonempty.
+   *
+   * @return potentially prioritized element from the head of the queue
+   */
+  private def dequeue: E = {
+    val e = if (prioritizedElements.isEmpty) regularElements.poll else prioritizedElements.poll
```

Comment:
> Hmm, when an item is added to an empty queue, there could be multiple request handler threads calling poll(). They could all get notified at the same time and all see prioritizedElements.isEmpty being false and one of them will then be blocked forever. One implication of this is that the requestHandlerIdleAvg metric won't be updated.

This doesn't sound right for multiple reasons:
1. A put only calls signal, not signalAll, so only a single waiting thread is woken up. Note that the code would still work even if signalAll were to be called.
2. Regardless of the signal variant called, the thread(s) that are woken up re-acquire the lock before returning from await. This means only one thread is active in the critical section with the lock at any point in time. Access to all of our internal data structures are synchronized under the lock.
3. RequestChannel today already uses ArrayBlockingQueue. ArrayBlockingQueue's put API will block indefinitely until it's interrupted or no longer full. The PrioritizationAwareBlockingQueue behaves the same way. You can compare ArrayBlockingQueue's implementation with PrioritizationAwareBlockingQueue if you want.

Reply:
Sorry, didn't see the lock. So, this is fine then.

Reply:
@junrao Good point about KIP-72. If we follow the original idea to separate control plain and data plain. The controller to broker connection should have a dedicated selector which should never get throttled. It is likely OK because we know that at any given point there should only be one controller and at most one controller request in process.